File: java_programming/MergeSort.java
import java.util.Scanner;
import java.util.Random;
public class Program5 {
	public static void main(String[] args)
	{
		Scanner read= new Scanner(System.in);
		int a[]= new int[10000];
		long start, end; int i;
		System.out.println("Enter the number of elements to sort");
		int n=read.nextInt();
		Random rand= new Random();
		for(i=0;i<n;i++)
			a[i]=rand.nextInt(100);
		System.out.println("Array elements to be sorted are");
		for(i=0;i<n;i++)
			System.out.print(a[i]+" ");
		System.out.println();
		//Program5 ob=new Program5();
		start=System.nanoTime();
		MergeSort(a,0,n-1);
		end=System.nanoTime();
		System.out.println("The sorted array is");
		for(i=0;i<n;i++)
			System.out.print(a[i]+" ");
		System.out.println();
		System.out.println("The time taken "+(end-start)+"ns");  
	}
    static void MergeSort(int a[], int l, int h)
    {
    	int mid;
    	if(l<h)
    	{
    		mid=(l+h)/2;
    		MergeSort(a,l,mid);
    		MergeSort(a,mid+1,h);
    		Merge(a,l,mid,h);
    	}
    }
    static void Merge(int a[], int l, int mid, int h)
    {
    	int i=l,j=mid+1,k=l,m,b[]=new int[10000];
    	while((i<=mid)&&(j<=h))
    	{
    		if(a[i]<=a[j])
    			b[k++]=a[i++];
    		else
    			b[k++]=a[j++];
    	}
    	if(i>mid)
    	  for(m=j;m<=h;m++)
    		  b[k++]=a[m];
    	else
    		for(m=i;m<=mid;m++)
    			b[k++]=a[m];
    	for(m=l;m<=h;m++)
    		a[m]=b[m];
    }
}

File: Hacktoberfest_2020/types_of_sorting/Rishikesh_Kumar/merge_sort.cpp
#include <iostream>
using namespace std;

void merge(int arr[], int l, int mid, int r)
{
    int n1 = mid - l + 1;
    int n2 = r - mid;
    int a[n1], b[n2];

    for (int i = 0; i < n1; i++) a[i] = arr[l+i];
    for (int i = 0; i < n2; i++) b[i] = arr[mid+i+1];


    int i = 0, j = 0, k = 0;

    while (i < n1 && j < n2)
    {
        if (a[i] >= b[j])
        {
            arr[l+k] = b[j];
            j++;
        }
        else
        {
            arr[l+k] = a[i];
            i++;
        }
        k++;
    }

    while (i < n1) 
    {
        arr[l+k] = a[i];
        i++; k++;
    }

    while (j < n2) 
    {
        arr[l+k] = b[j];
        k++; j++;
    }
}

void mergeSort(int arr[], int l, int r)
{
    if (l >= r) return ;

    int mid = (l+r)/2;

    mergeSort(arr, l, mid);
    mergeSort(arr, mid+1, r);
    merge(arr, l, mid, r);

    return;
}

int main()
{

    int n;
    cout << "Enter the number of elements: ";
    cin >> n;

    cout << "Enter the elements: ";
    int arr[n];

    for (int i = 0; i < n; i++) cin >> arr[i];

    mergeSort(arr, 0, n-1);

    cout << "The sorted elements are: ";
    for (int i = 0; i < n; i++) cout << arr[i] << " ";
    cout << endl;

    return 0;
}